python/numpy/weldarray.py

```diff
@@ -0,0 +1,375 @@
+            return WeldArray(result)
+        else:
+            return result
+
+    def evaluate(self):
+        '''
+        User facing function - if he wants to explicitly evaluate all the
+        registered ops.
+        '''
+        return WeldArray(self._eval())
+
+    def _eval(self):
+        '''
+        @ret: ndarray after evaluating all the ops registered with the given
+        WeldArray.
+
+        Users should instead call evaluate which would return a WeldArray as
+        expected. This returns an ndarray.
+
+        Evalutes the expression based on weldobj.weld_code. If no new ops have
+        been registered, then just returns the last ndarray.
+        '''
+        if self.name == self.weldobj.weld_code:
+            # no new ops have been registered. Let's avoid creating
+            # unneccessary new copies with weldobj.evaluate()
+            arr = self.weldobj.context[self.name]
+            return arr
+
+        arr = self.weldobj.evaluate(WeldVec(self._weld_type),
+                verbose=self.verbose)
+        # Now that the evaluation is done - create new weldobject for self,
+        # initalized from the given arr.
+        self.weldobj = WeldObject(NumpyArrayEncoder(), NumpyArrayDecoder())
+        self.weldobj.weld_code = ''
+        self.name = self.weldobj.update(arr, SUPPORTED_DTYPES[str(arr.dtype)])
+        self.weldobj.weld_code = self.name
+
+        return arr
+
+    def _unary_op(self, unop, result=None):
+        '''
+        Create a new array, and updates weldobj.weld_code with the code for
+        unop.
+        @unop: str, weld IR for the given function.
+        @result: output array.
+        '''
+        def _update_array_unary_op(res, unop):
+            '''
+            '''
+            template = 'map({arr}, |z : {type}| {unop}(z))'
+            code = template.format(arr = res.weldobj.weld_code,
+                    type=res._weld_type.__str__(),
+                    unop=unop)
+            res.weldobj.weld_code = code
+
+        if result is None:
+            result = WeldArray(self)
+        else:
+            # if array had view / was a view - need to update the other
+            # appropriate arrays too.
+            for c in self.view_children:
+                _update_array_unary_op(c, unop)
+
+            # TODO: Updating parents.
+
+        _update_array_unary_op(result, unop)
+        return result
+
+    def _scalar_binary_op(self, other, binop, result):
+        '''
+        Helper function for _binary_op.
+        @other is a scalar (i32, i64, f32, f64).
+        @result: None or a weldarray to store results in.
+        '''
+        if result is None:
+            result = WeldArray(self)
+        else:
+            # FIXME: deal with updating view children/parents
+            pass
+
+        # FIXME: support other dtypes - not sure what to do for f64/or i64.
+        if isinstance(other, np.float32) or isinstance(other, float):
```

Comment:
Better way to do this might be to have some dictionary called `DTYPE_SUFFIXES`, which maps every value in `SUPPORTED_DTYPES` to a string (e.g., `""` for `i32`, `"f"` for `f32`, etc.). That way, when we add new types we don't have to go through all the code and find the points where support for the type needs to be added (then here, you would just index into the dictionary with the `dtype`.

You could even make a class called `DTypeInfo` or something and make this suffix a field in that object, so we need to add other type specific stuff we can append things to that object.